chaos_genius/views/download_view.py

```diff
@@ -0,0 +1,215 @@
+        ] = f"attachment; filename=KPI-{kpi_id}-anomaly-data.csv"
+        return response
+    except Exception as e:
+        return jsonify(
+            {"status": "failure", "message": f"Downloading data failed: {e}"}
+        )
+
+
+@blueprint.route("/<int:kpi>/chart_data", methods=["GET"])
+def kpi_download_line_data(kpi: int):
+    """API endpoint to download chart data for last 60 days."""
+    try:
+        line_data = kpi_line_data(kpi, download=True)
+        if line_data[0] == "error":
+            raise Exception(f"{line_data[1]}")
+        data_points = line_data[2]
+
+        if not data_points:
+            raise Exception(f"No chart data found for KPI id {kpi}")
+
+        def row_gen(data_points):
+            csv_headers = ["date", "value"]
+            yield csv_headers
+            for row in data_points:
+                attr_list = [row["date"].strftime("%a %-d %B %Y"), str(row["value"])]
+                yield attr_list
+
+        response = Response(iter_csv(row_gen(data_points)), mimetype="text/csv")
+        response.headers[
+            "Content-Disposition"
+        ] = f"attachment; filename=KPI-{kpi}-panel-chart-data.csv"
+        return response
+    except Exception as e:
+        return jsonify(
+            {"status": "failure", "message": f"chart data download failed: {e}"}
+        )
+
+
+@blueprint.route("/<int:kpi_id>/hierarchical_data", methods=["GET"])
+def download_hierarchical_data(kpi_id):
+    """API endpoint to download RCA hierarchical data."""
+    try:
+        timeline = request.args.get("timeline")
+        if not timeline:
+            raise Exception("Please provide timeline as an argument")
+
+        kpi_info = get_kpi_data_from_id(kpi_id)
+        dimensions = kpi_info.get("dimensions")
+        if not dimensions:
+            raise Exception(f"Failed to fetch dimensions for kpi {kpi_id}")
+
+        data_list = {}
+        for dimension in dimensions:
+            result = rca_hierarchical_data(kpi_id, timeline, dimension)
+            if result[0] == "error":
+                raise Exception(
+                    f"failed to fetch heirarchial data for kpi {kpi_id} with dimension={dimension} - {result[1]}"
+                )
+            data_list[dimension] = result[2]["data_table"]
+
+        def row_gen(data_list):
+            csv_headers = [
+                "Dimension Name",
+                "Hierarchy Level",
+                "Sub-Group Name",
+                "Previous Period Value",
+                "Previous Period Count(#)",
+                "Previous Period Size(%)",
+                "Current Period Value",
+                "Current Period Count(#)",
+                "Current Period Size(%)",
+                "Impact",
+            ]
+            yield csv_headers
+
+            for dimension in data_list.keys():
+                data = data_list[dimension]
+                for row in data:
+                    hlevel = len(row["subgroup"].split("&"))
```

Comment:
Yes this is an edge case, for now it is handled by splitting it on ' & ' since the hierarchical level is not stored. It can be calculated but would be too inefficient to do this when generating the CSV. It would be best to include hlevel when inserting hierarchical data into the DB.